File: apps/ios/App/Auth/AuthViewModel.swift
import Foundation
import SwiftUI
import Factory
import Data

@Observable
@MainActor
public final class AuthViewModel {
    private let authRepository: AuthRepository
    
    public private(set) var authState: AuthState = .signedOut
    public private(set) var isLoading: Bool = false
    public private(set) var error: String? = nil
    public var email: String = ""
    public var password: String = ""
    public var otp: String = ""
    public var otpSent: Bool = false

    /// Which way in the user is currently using. Matches Android's
    /// `LoginScreen.Mode` so the two screens offer the same two doors.
    public enum Mode: Sendable { case password, otp }
    public var mode: Mode = .password

    /// The three steps of a password reset, plus "not resetting".
    ///
    /// A separate axis from `otpSent`, not a fifth value of it: sign-in-by-OTP
    /// and reset-by-OTP both show a 6-digit field, but they verify with
    /// DIFFERENT OTP types and end somewhere different. Collapsing them into
    /// one flag is how the wrong type gets used.
    public enum ResetStage: Sendable { case none, email, code, newPassword }
    public var resetStage: ResetStage = .none
    public var confirmPassword: String = ""
    
    public init(authRepository: AuthRepository) {
        self.authRepository = authRepository
        Task {
            for await state in authRepository.authState {
                await MainActor.run {
                    self.authState = state
                }
            }
        }
    }
    
    @MainActor
    public func continueAsGuest() async {
        isLoading = true
        error = nil
        do {
            _ = try await authRepository.ensureUser()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
    
    @MainActor
    public func sendOtp() async {
        guard !email.isEmpty else { return }
        isLoading = true
        error = nil
        do {
            try await authRepository.sendOtp(email: email)
            otpSent = true
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
    
    @MainActor
    public func verifyOtpAndSignIn() async {
        guard !email.isEmpty, !otp.isEmpty else { return }
        isLoading = true
        error = nil
        do {
            try await authRepository.verifyOtp(email: email, token: otp)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
    
    @MainActor
    public func signInWithPassword() async {
        guard !email.isEmpty, !password.isEmpty else { return }
        isLoading = true
        error = nil
        do {
            try await authRepository.signInWithPassword(email: email, password: password)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    /// Continue with Google.
    ///
    /// `isLoading` stays true across the whole call because
    /// `ASWebAuthenticationSession` is presented on top of this screen and the
    /// call does not return until it is dismissed — unlike Android, where the
    /// browser is a separate task and the result comes back through a deep
    /// link.
    ///
    /// A user cancellation arrives here as a thrown error like any other. It is
    /// reported rather than swallowed: silently returning to an unchanged
    /// screen after a browser flashed past reads as a broken button.
    @MainActor
    public func continueWithGoogle() async {
        isLoading = true
        error = nil
        do {
            try await authRepository.continueWithGoogle()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    /// Enter the reset flow from the sign-in screen.
    @MainActor
    public func startPasswordReset() {
        resetStage = .email
        error = nil
    }

    /// Leave it — back to ordinary sign-in.
    @MainActor
    public func cancelPasswordReset() {
        resetStage = .none
        error = nil
    }

    /// Step 1 — send the code. Advances even though the call cannot tell us
    /// whether the account exists; see Auth.swift on why it must not.
    @MainActor
    public func sendPasswordReset() async {
        guard !email.isEmpty else { return }
        await run { try await self.authRepository.sendPasswordReset(email: self.email) } onSuccess: {
            self.resetStage = .code
        }
    }

    /// Step 2 — verify. Success leaves a short-lived recovery session.
    @MainActor
    public func verifyPasswordResetCode() async {
        guard !otp.isEmpty else { return }
        await run {
            try await self.authRepository.verifyPasswordResetCode(email: self.email, token: self.otp)
        } onSuccess: {
            self.resetStage = .newPassword
        }
    }

    /// Step 3 — set it. The auth gate takes over once the session is live.
    @MainActor
    public func setNewPassword() async {
        guard password.count >= Self.minPasswordLength, password == confirmPassword else { return }
        await run { try await self.authRepository.setPassword(self.password) } onSuccess: {
            self.resetStage = .none
        }
    }

    /// Web's `password.length < 8` check, named rather than inline.
    public static let minPasswordLength = 8

    /// One place that sets `isLoading`, clears the error, and reports failure.
    ///
    /// The four reset methods would otherwise repeat the same six lines each,
    /// which is how one of them ends up not clearing `error` and showing a
    /// stale message on the next step.
    @MainActor
    private func run(
        _ block: () async throws -> Void,
        onSuccess: () -> Void
    ) async {
        isLoading = true
        error = nil
        do {
            try await block()
            onSuccess()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    @MainActor
    public func signOut() async {
        isLoading = true
        do {
            try await authRepository.signOut()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }
}

File: apps/android/app/src/main/java/com/sanvya/app/ui/auth/AuthViewModel.kt
package com.sanvya.app.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.auth.AuthRepository
import com.sanvya.app.data.auth.AuthState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * Sign-in state for `LoginScreen`.
 *
 * **Every one of these used to end in `catch (e: Exception) { // Handle error }`.**
 * A failed OTP send did nothing at all: no message, and the success callback
 * never fired, so the screen simply did not advance and the user had no idea
 * why. Errors are surfaced now — a swallowed error is the same failure as a
 * dead button, one layer down.
 */
class AuthViewModel : ViewModel(), KoinComponent {
    // `by inject()` rather than constructor injection: no view model in this app
    // is registered with Koin, and `viewModel()` needs a no-arg constructor.
    // Matching AccountsViewModel and the rest rather than introducing a second
    // wiring style for one screen.
    private val authRepository: AuthRepository by inject()

    val authState: StateFlow<AuthState> = authRepository.authState.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = AuthState.SIGNED_OUT
    )

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _busy = MutableStateFlow(false)
    val busy: StateFlow<Boolean> = _busy.asStateFlow()

    /**
     * Where the password-reset flow has got to, if it has started.
     *
     * A separate axis from [otpSent], not a fifth value of it: sign-in-by-OTP
     * and reset-by-OTP both show a 6-digit code field, but they verify with
     * DIFFERENT OTP types and end somewhere different. Collapsing them into one
     * flag is how the wrong type gets used.
     */
    private val _resetStage = MutableStateFlow(ResetStage.NONE)
    val resetStage: StateFlow<ResetStage> = _resetStage.asStateFlow()

    /** True once a code has been sent, so the screen shows the code step. */
    private val _otpSent = MutableStateFlow(false)
    val otpSent: StateFlow<Boolean> = _otpSent.asStateFlow()

    fun clearError() { _error.value = null }

    /** Enter the reset flow from the sign-in screen. */
    fun startPasswordReset() {
        _resetStage.value = ResetStage.EMAIL
        _error.value = null
    }

    /** Leave it — back to ordinary sign-in. */
    fun cancelPasswordReset() {
        _resetStage.value = ResetStage.NONE
        _error.value = null
    }

    /** Step 1 — send the code. Advances even though the call cannot tell us
     *  whether the account exists; see Auth.kt on why it must not. */
    fun sendPasswordReset(email: String) =
        run({ _resetStage.value = ResetStage.CODE }) { authRepository.sendPasswordReset(email) }

    /** Step 2 — verify. Success leaves a short-lived recovery session. */
    fun verifyPasswordResetCode(email: String, token: String) =
        run({ _resetStage.value = ResetStage.NEW_PASSWORD }) {
            authRepository.verifyPasswordResetCode(email, token)
        }

    /** Step 3 — set it. The auth gate takes over once the session is live. */
    fun setPassword(password: String, onDone: () -> Unit = {}) =
        run({ _resetStage.value = ResetStage.NONE; onDone() }) {
            authRepository.setPassword(password)
        }

    fun backToEmail() {
        _otpSent.value = false
        _error.value = null
    }

    /**
     * Runs one auth call, reporting failure rather than hiding it.
     *
     * `onSuccess` fires ONLY on success. The old code called its callback
     * inside the `try` after the await, which was right, but the empty `catch`
     * meant a failure was indistinguishable from a call that never happened.
     */
    private fun run(onSuccess: () -> Unit = {}, block: suspend () -> Unit) {
        viewModelScope.launch {
            _busy.value = true
            _error.value = null
            try {
                block()
                onSuccess()
            } catch (e: Exception) {
                _error.value = e.message ?: "Something went wrong. Please try again."
            } finally {
                _busy.value = false
            }
        }
    }

    fun ensureGuest(onComplete: () -> Unit = {}) = run(onComplete) { authRepository.ensureGuest() }

    fun sendOtp(email: String) = run({ _otpSent.value = true }) { authRepository.sendOtp(email) }

    fun verifyOtp(email: String, token: String, onVerified: () -> Unit = {}) =
        run(onVerified) { authRepository.verifyOtp(email, token) }

    fun signUp(email: String, password: String, username: String, onComplete: () -> Unit = {}) =
        run(onComplete) { authRepository.signUp(email, password, username) }

    fun signInWithPassword(email: String, password: String, onComplete: () -> Unit = {}) =
        run(onComplete) { authRepository.signInWithPassword(email, password) }

    /**
     * Continue with Google.
     *
     * Deliberately takes no success callback. Everything else here finishes
     * when the suspend call returns; this one only *launches a browser*, and
     * the session appears later when the OS routes the callback URI back into
     * MainActivity. A callback would fire the moment the Custom Tab opened,
     * which is precisely the kind of control that looks like it worked and did
     * not.
     *
     * `busy` is not cleared on success either: the app is going to the
     * background, and a spinner that keeps spinning until the callback lands is
     * the honest state. It clears on failure, which is when the user is still
     * looking at this screen.
     */
    fun continueWithGoogle() = run { authRepository.continueWithGoogle() }

    /** The three steps of a password reset, plus "not resetting". */
    enum class ResetStage { NONE, EMAIL, CODE, NEW_PASSWORD }
}
